### PR TITLE
config: add pd client option to enable request redirection

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Problem Summary:

When there is a network partition problem happens between TiDB and PD or allocator leader, it can not serve anymore at the moment.

### What is changed and how it works?

What's Changed:

How it Works: This PR is going to add an option to make PD client can redirect the requests to PD or allocator leader if there is a network partition problem.

This PR should be merged after https://github.com/tikv/pd/pull/3431 is merged.

Tests 

- Manual test

### Release note 

- Add PD client option to enable request redirection

### CODE REVIEW

File: store/driver/tikv_driver.go [hunk 1]
```diff
@@ -121,12 +121,15 @@ func (d TiKVDriver) OpenWithOptions(path string, options ...Option) (kv.Storage,
...
-	), pd.WithCustomTimeoutOption(time.Duration(d.pdConfig.PDServerTimeout)*time.Second))
+	},
+		pd.WithGRPCDialOptions(
+			grpc.WithKeepaliveParams(keepalive.ClientParameters{
+				Time:    time.Duration(d.tikvConfig.GrpcKeepAliveTime) * time.Second,
+				Timeout: time.Duration(d.tikvConfig.GrpcKeepAliveTimeout) * time.Second,
+			}),
+		),
+		pd.WithCustomTimeoutOption(time.Duration(d.pdConfig.PDServerTimeout)*time.Second),
+		pd.WithForwardingOption(d.pdConfig.EnableForwarding))
```
Comment (BusyJay) [inline store/driver/tikv_driver.go:132]: How about making it a global config so tikv client can also check the option?

File: config/config.go [hunk 2]
```diff
@@ -176,6 +176,9 @@ type Config struct {
 	// EnableTCP4Only enables net.Listen("tcp4",...)
 	// Note that: it can make lvs with toa work and thus tidb can get real client ip.
 	EnableTCP4Only bool `toml:"enable-tcp4-only" json:"enable-tcp4-only"`
+	// The client will forward the requests to the PD or allocator leader through the follower
```
Comment (BusyJay) [inline config/config.go:179]: The comment should be updated to be more generic as it doesn't just control the behavior of PD client.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

File: config/config.go (hunk 2 above)
Comment (rleungx) [reply config/config.go:179]: Oh, I forgot about it.

**Comment by rleungx** [comment]
UnlhbiBMZXVuZw==

**Review by nolouch — APPROVED** [review]
lgtm

**Review by nolouch — APPROVED** [review]

**Review by BusyJay — APPROVED** [review]

File: config/config_test.go [hunk 3]
```diff
@@ -278,6 +278,7 @@ spilled-file-encryption-method = "plaintext"
 	c.Assert(conf.Security.SpilledFileEncryptionMethod, Equals, SpilledFileEncryptionMethodPlaintext)
 	c.Assert(conf.DeprecateIntegerDisplayWidth, Equals, true)
 	c.Assert(conf.EnableEnumLengthLimit, Equals, false)
+	c.Assert(conf.EnableForwarding, Equals, false)
```
Comment (MyonKeminta) [inline config/config_test.go:281]: In this test, you should try to set the value to a non-default value in the config text above, and assert it's actually set to a non-default value. So you should add `enable-forwarding=true` at about L200 and assert it's true here.

File: store/driver/tikv_driver.go [hunk 4]
```diff
@@ -24,6 +24,7 @@ import (
 	"time"
 
 	"github.com/pingcap/errors"
+	tidbconfig "github.com/pingcap/tidb/config"
```
Comment (MyonKeminta) [inline store/driver/tikv_driver.go:27]: I'm afraid that you may need to check if it breaks the work about extracting the tikvclient out from tidb repo.
cc @disksing

**Review by MyonKeminta — COMMENTED** [review]
Rest LGTM

**Review by MyonKeminta — APPROVED** [review]
LGTM

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**